core/.js/src/main/scala/io/udash/component/ComponentId.scala

```diff
@@ -1,20 +1,38 @@
 package io.udash.component
 
-/** General usage class for carrying id of component. */
-case class ComponentId(id: String) extends AnyVal {
-  override def toString: String = id
+import com.avsystem.commons.misc.CaseMethods
+import io.udash.macros.ComponentIdMacro
+import org.scalajs.dom.Element
+import scalatags.JsDom.GenericAttr
+import scalatags.JsDom.all._
 
-  def subcomponent(subId: String): ComponentId =
-    ComponentId(s"$id-$subId")
+
+final case class ComponentId private(value: String) extends Modifier with CaseMethods {
+  override def applyTo(t: Element): Unit = {
+    t.id = value
+  }
+
+  /** Generate new [[ComponentId]] based on a current id value with added suffix. */
+  def withSuffix(s: String): ComponentId = ComponentId.forName(ComponentId.join(value, s))
+
+  override def toString(): String = value
 }
 
 object ComponentId {
-  private var cid = -1
+  final val Separator = "-"
```

Comment:
The `ComponentId` doesn't preserve the parts though but it may not be possible to recover them from the name even with public `Separator` (consider numerical suffices). Let's try with private join as well then.